From: Peter Michael Green <plugwash@debian.org>
Date: Tue, 12 Nov 2019 12:54:02 +0100
Subject: Attempt at opencv 4 support.

  [Peter Michael Green]
    + Bump libopencv-dev build-dependency as I am not attempting to retain
      compatibility with older opencv versions, someone who knows the
      buildsystem better can do that if they wish.
    + Change pkg-config file for opencv from "opencv" to "opencv4".
    + Change/add a bunch of #include statements

  [Ichthyostega]
    + adapt this patch to work with more recent GMic/CImg (2.8.4)
---
 CMakeLists.txt                 | 2 +-
 src/CImg.h                     | 3 +++
 src/Makefile                   | 4 ++--
 zart/include/ImageConverter.h  | 1 +
 zart/include/VideoFileSource.h | 2 ++
 zart/zart.pro                  | 2 +-
 6 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index ffbb12e..c1aad68 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -184,7 +184,7 @@ if(FFTW3_FOUND)
 endif()
 
 if(ENABLE_OPENCV)
-  pkg_check_modules(OPENCV opencv)
+  pkg_check_modules(OPENCV opencv4)
 endif()
 if(OPENCV_FOUND)
   set(CLI_COMPILE_FLAGS "${CLI_COMPILE_FLAGS} -Dcimg_use_opencv")
diff --git a/src/CImg.h b/src/CImg.h
index 6c079ff..3f15392 100644
--- a/src/CImg.h
+++ b/src/CImg.h
@@ -438,7 +438,10 @@ enum {FALSE_WIN = FALSE};
 #define _cimg_cap_prop_frame_height CV_CAP_PROP_FRAME_HEIGHT
 #define _cimg_cap_prop_frame_count CV_CAP_PROP_FRAME_COUNT
 #endif
+#if CV_MAJOR_VERSION>=4
+#include "opencv2/videoio/videoio_c.h"
 #endif
+#endif // cimg_use_opencv
 
 // Configure LibPNG support.
 // (http://www.libpng.org)
diff --git a/src/Makefile b/src/Makefile
index 4a5a3c2..54b1ae6 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -289,8 +289,8 @@ CURL_LIBS = $(shell pkg-config --libs libcurl || echo -lcurl)
 
 # Enable native support of webcams and video streaming, using the OpenCV library.
 # (https://opencv.org/)
-OPENCV_CFLAGS = -Dcimg_use_opencv $(shell pkg-config opencv --cflags || echo -I/usr/include/opencv) -I/usr/include/opencv
-OPENCV_LIBS = $(shell pkg-config opencv --libs || echo -lopencv_core -lopencv_highgui)
+OPENCV_CFLAGS = -Dcimg_use_opencv $(shell pkg-config opencv4 --cflags || echo -I/usr/include/opencv) -I/usr/include/opencv
+OPENCV_LIBS = $(shell pkg-config opencv4 --libs || echo -lopencv_core -lopencv_highgui)
 
 # Enable support of most classical image file formats, using the GraphicsMagick++ library.
 # (http://www.graphicsmagick.org/Magick++/)
diff --git a/zart/include/ImageConverter.h b/zart/include/ImageConverter.h
index 42e907b..7c690fa 100644
--- a/zart/include/ImageConverter.h
+++ b/zart/include/ImageConverter.h
@@ -48,6 +48,7 @@
 
 #include <QMutex>
 #include <opencv2/opencv.hpp>
+#include <opencv2/imgproc/imgproc_c.h>
 #ifndef gmic_build
 #include "CImg.h"
 #endif
diff --git a/zart/include/VideoFileSource.h b/zart/include/VideoFileSource.h
index bd45c33..dac345c 100644
--- a/zart/include/VideoFileSource.h
+++ b/zart/include/VideoFileSource.h
@@ -49,6 +49,8 @@
 
 #include <QString>
 #include <opencv2/opencv.hpp>
+#include <opencv2/videoio/videoio_c.h>
+#include <opencv2/highgui/highgui_c.h>
 #include "ImageSource.h"
 
 class VideoFileSource : public ImageSource {
diff --git a/zart/zart.pro b/zart/zart.pro
index 7bc66a8..e884a8e 100644
--- a/zart/zart.pro
+++ b/zart/zart.pro
@@ -29,7 +29,7 @@ greaterThan(QT_MAJOR_VERSION, 4): CONFIG += c++11
 CONFIG	+= warn_on
 QT_CONFIG -= no-pkg-config
 CONFIG += link_pkgconfig
-PKGCONFIG += opencv fftw3 zlib
+PKGCONFIG += opencv4 fftw3 zlib
 DEFINES += cimg_use_fftw3 cimg_use_zlib
 
 DEFINES += QT_DEPRECATED_WARNINGS
